compiler-core/templates/prelude.mjs

```diff
@@ -156,38 +156,76 @@ export function toBitArray(segments) {
 
 // @internal
 // Derived from this answer https://stackoverflow.com/questions/8482309/converting-javascript-integer-to-byte-array-and-back
-export function sizedInt(int, size) {
-  let value = int;
+export function sizedInt(value, size, isBigEndian) {
   if (size < 0) {
     return new Uint8Array();
   }
-  if (size % 8 != 0) {
-    const msg = `Bit arrays must be byte aligned on JavaScript, got size of ${size} bits`;
-    throw new globalThis.Error(msg);
-  }
+
   const byteArray = new Uint8Array(size / 8);
 
-  for (let index = 0; index < byteArray.length; index++) {
-    const byte = value & 0xff;
-    byteArray[index] = byte;
-    value = (value - byte) / 256;
+  // Convert negative number to two's complement representation
+  if (value < 0) {
+    value = (2 ** size) + value;
+  }
+
+  if (isBigEndian) {
+    for (let i = 0; i < byteArray.length; i++) {
+      const byte = value % 256
+      byteArray[i] = byte;
+      value = (value - byte) / 256;
+    }
+  } else {
+    for (let i = byteArray.length - 1; i >= 0; i--) {
+      const byte = value % 256
+      byteArray[i] = byte;
+      value = (value - byte) / 256;
+    }
   }
+
   return byteArray.reverse();
 }
 
 // @internal
-export function byteArrayToInt(byteArray) {
-  byteArray = byteArray.reverse();
+export function byteArrayToInt(byteArray, start, end, isBigEndian, isSigned) {
   let value = 0;
-  for (let i = byteArray.length - 1; i >= 0; i--) {
-    value = value * 256 + byteArray[i];
+
+  // Read bytes as an unsigned integer value
+  if (isBigEndian) {
+    for (let i = start; i < end; i++) {
+      value = value * 256 + byteArray[i];
+    }
+  } else {
+    for (let i = end - 1; i >= start; i--) {
+      value = value * 256 + byteArray[i];
+    }
   }
+
+  if (isSigned) {
+    const byteSize = end - start;
+
+    const highBit = 2 ** (byteSize * 8 - 1);
+
+    // If the high bit is set and this is a signed integer, so reinterpret
+    // as two's complement
+    if (value >= highBit) {
+      value -= highBit * 2;
+    }
+  }
+
   return value;
 }
 
 // @internal
-export function byteArrayToFloat(byteArray) {
-  return new Float64Array(byteArray.reverse().buffer)[0];
+export function byteArrayToFloat(byteArray, start, end, isBigEndian) {
+  const view = new DataView(byteArray.buffer);
+
+  const byteSize = end - start;
+
+  if (byteSize === 8) {
+    return view.getFloat64(start, !isBigEndian)
+  } else if (byteSize === 4) {
+    return view.getFloat32(start, !isBigEndian)
+  }
```

Comment:
This function returns `undefined` when byteSize is not 8 or 4. It must always return a valid value.

Rather than have an end we could have a second bool for whether it is 32 or 64 bits.

Reply:
This function is currently invalid and can return undefined, it must be fixed. No code should depend on always being used correctly as then mistakes can result in cryptic and hard to debug incorrect behaviour instead of an easy to fix error.

Reply:
For sure, did you see the throw that was added to this function last month to avoid returning undefined?

<img width="812" alt="image" src="https://github.com/user-attachments/assets/c195c3d1-937e-4e51-a90c-33e98304d71e">

Reply:
Oh sorry, I did not. Thank you